#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: fix insecure temporary file handling in glibcbug, quote all
# DP: $TEMP and $TEMPX variables
# DP: Related bugs: Debian bug #278278, Warty bug #2743
# DP: Upstream status: CVS removed glibcbug, thus not vulnerable

PATCHLEVEL=1

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p$PATCHLEVEL < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p$PATCHLEVEL < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.
--- glibc-2.3.2/glibcbug.in
+++ glibc-2.3.2.new/glibcbug.in
@@ -22,18 +22,14 @@
 BUILD_BOUNDED="@bounded@"
 BUILD_STATIC_NSS="@static_nss@"
 
-TEMP=`mktemp -q ${TMPDIR-/tmp}/glibcbugXXXXXX 2>/dev/null`
-if test $? -ne 0; then
-  TEMP=${TMPDIR-/tmp}/glibcbug.$$
-  echo > $TEMP
-  chmod 600 $TEMP
-fi
-TEMPx=`mktemp -q ${TMPDIR-/tmp}/glibcbugXXXXXX 2>/dev/null`
-if test $? -ne 0; then
-  TEMPx=${TMPDIR-/tmp}/glibcbug.$$.x
-  echo > $TEMPx
-  chmod 600 $TEMPx
-fi
+TEMP="`mktemp -t glibcbugXXXXXXXXXX`" || exit 1
+TEMPx="`mktemp -t glibcbugXXXXXXXXXX`" || {
+  rm -f "$TEMP"
+  exit 1
+}
+
+trap 'rm -f "$TEMP" "$TEMPx"; exit 1' HUP INT PIPE TERM
+trap 'rm -f "$TEMP" "$TEMPx"' EXIT
 
 BUGGLIBC="glibc-bug-reports-${RELEASE}@gnu.org"
 BUGADDR=${1-$BUGGLIBC}
@@ -42,10 +38,6 @@
 
 : ${USER=${LOGNAME-`whoami`}}
 
-trap 'rm -f $TEMP $TEMPx; exit 1' 1 2 3 13 15
-trap 'rm -f $TEMP $TEMPx' 0
-
-
 # How to read the passwd database.
 PASSWD="cat /etc/passwd"
 
@@ -72,8 +64,8 @@
 else
   # Must use temp file due to incompatibilities in quoting behavior
   # and to protect shell metacharacters in the expansion of $LOGNAME
-  $PASSWD | grep "^$LOGNAME:" | awk -F: '{print $5}' | sed -e 's/,.*//' > $TEMP
-  ORIGINATOR="`cat $TEMP`"
+  $PASSWD | grep "^$LOGNAME:" | awk -F: '{print $5}' | sed -e 's/,.*//' > "$TEMP"
+  ORIGINATOR="`cat "$TEMP"`"
 fi
 
 if [ -n "$ORGANIZATION" ]; then
@@ -123,7 +115,7 @@
 FIX_C='<how to correct or work around the problem, if known (multiple lines)>'
 
 
-cat > $TEMP <<EOF
+cat > "$TEMP" <<EOF
 SEND-PR: -*- send-pr -*-
 SEND-PR: Lines starting with \`SEND-PR' will be removed automatically, as
 SEND-PR: will all comments (text enclosed in \`<' and \`>').
@@ -171,12 +163,12 @@
 	$FIX_C
 EOF
 
-chmod u+w $TEMP
-cp $TEMP $TEMPx
+chmod u+w "$TEMP"
+cp "$TEMP" "$TEMPx"
 
-eval $EDIT $TEMP
+eval $EDIT "$TEMP"
 
-if cmp -s $TEMP $TEMPx; then
+if cmp -s "$TEMP" "$TEMPx"; then
 	echo "File not changed, no bug report submitted."
 	exit 1
 fi
@@ -205,7 +197,7 @@
   # 1) Severity
   #
   PATTERN=">Severity:"
-  SEVERITY=`eval sed -n -e "\"$SED_CMD\"" $TEMP`
+  SEVERITY=`eval sed -n -e "\"$SED_CMD\"" "$TEMP"`
   case "$SEVERITY" in
     ""|non-critical|serious|critical) CNT=`expr $CNT + 1` ;;
     *)  echo "$COMMAND: \`$SEVERITY' is not a valid value for \`Severity'."
@@ -214,7 +206,7 @@
   # 2) Priority
   #
   PATTERN=">Priority:"
-  PRIORITY=`eval sed -n -e "\"$SED_CMD\"" $TEMP`
+  PRIORITY=`eval sed -n -e "\"$SED_CMD\"" "$TEMP"`
   case "$PRIORITY" in
     ""|low|medium|high) CNT=`expr $CNT + 1` ;;
     *)  echo "$COMMAND: \`$PRIORITY' is not a valid value for \`Priority'."
@@ -223,7 +215,7 @@
   # 3) Class
   #
   PATTERN=">Class:"
-  CLASS=`eval sed -n -e "\"$SED_CMD\"" $TEMP`
+  CLASS=`eval sed -n -e "\"$SED_CMD\"" "$TEMP"`
   case "$CLASS" in
     ""|sw-bug|doc-bug|change-request|support) CNT=`expr $CNT + 1` ;;
     *)  echo "$COMMAND: \`$CLASS' is not a valid value for \`Class'."
@@ -238,11 +230,11 @@
     case "$input" in
       a*)
 	echo "$COMMAND: problem report saved in $HOME/dead.glibcbug."
-	cat $TEMP >> $HOME/dead.glibcbug
+	cat "$TEMP" >> $HOME/dead.glibcbug
         xs=1; exit
         ;;
       e*)
-        eval $EDIT $TEMP
+        eval $EDIT "$TEMP"
         continue 2
         ;;
       s*)
@@ -269,15 +261,15 @@
 /^>Description:/,/^>[A-Za-z-]*:/s;$DESCRIPTION_C;;
 /^>How-To-Repeat:/,/^>[A-Za-z-]*:/s;$HOW_TO_REPEAT_C;;
 /^>Fix:/,/^>[A-Za-z-]*:/s;$FIX_C;;
-" $TEMP > $TEMPx
+" "$TEMP" > "$TEMPx"
 
-if $MAIL_AGENT < $TEMPx; then
+if $MAIL_AGENT < "$TEMPx"; then
   echo "$COMMAND: problem report sent"
   xs=0; exit
 else
   echo "$COMMAND: mysterious mail failure, report not sent."
   echo "$COMMAND: problem report saved in $HOME/dead.glibcbug."
-  cat $TEMP >> $HOME/dead.glibcbug
+  cat "$TEMP" >> $HOME/dead.glibcbug
 fi
 
 exit 0
